### PR TITLE
Add "Available since version" to function reference

### LINKED ISSUE #417: add version provenance: (since [at least]) version x
#### Summary:

For the functions reference, add a provenance for each function indicating when it was introduced.   

#### Description:

This is especially important now that the interfaces are releasing asynchronously and users often have later doc than the version of Stan they're using.

#### Additional Information:

It will be OK to follow the Python docs and allow "since at least".

It would also be nice to add the same thing for language features, but it's less clear where to document them.  For example, the `complex` type was introduced recently, but types are described all over the doc.

There are lots of place this is listed: release notes, auto-generated function sigs, the actual stanc3 or previous stanc.

#### Current Version:
v2.28.0

### PR DESCRIPTION
**Pull request by WardBrian**
#### Submission Checklist

- [x] Builds locally 
- [x] Declare copyright holder and open-source license: see below

#### Summary

This closes #417. Each function has had a version number added to it. 

This represents the minimum version one should expect the function to appear in _in its current form_, meaning renames or changes in return type/signature count as moving up a version. 

This was built from a list automatically generated from old documentation back to version 2.0, with some manual adjustments provided by @rok-cesnovar, @bob-carpenter, and by looking through old release notes. I'm fairly confident most of the version numbers are correct, and when they miss they miss _newer_, so we never say something was available during a version it wasn't, we just may be late here or there.

Here is how it looks in PDF:

Screenshot from 2021-10-12 14-49-31

And web:

Screenshot from 2021-10-12 14-49-58

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):
Simons Foundation

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: src/functions-reference/real-valued_basic_functions.Rmd
Comment (bob-carpenter) [inline src/functions-reference/real-valued_basic_functions.Rmd:297]: [comment]
more good catches.  thanks!

File: src/functions-reference/unbounded_continuous_distributions.Rmd
Comment (bob-carpenter) [inline src/functions-reference/unbounded_continuous_distributions.Rmd:818]: [optional]
I strongly prefer`\mid` over `|` as it puts a bit of space around the vertical bar.

Also, it's conventional to have the start/end math symbols (here `$$`) be on their own line.

It's also sub-optimal to have lines wrap---better to break up expressions into natural chunks.

The use of `I` should be `\textrm{I}` because that's how special functions get set in math.

I prefer to use `(`  everywhere instead of using `{` or `[` in some places.  If they get confusing to parse, either think about rewriting in smaller chunks or using `\left(` and `\right)`.

**Review by bob-carpenter — CHANGES_REQUESTED** [review]
A question, some optional fixes depending on your patience, and a lot of gratitude for finding and fixing so many existing errors.  This doesn't necessarily need any fixes, but I'm still marking as requesting changes to wait for answers to the questions before merging.

File: src/functions-reference/complex-valued_basic_functions.Rmd
Comment (WardBrian) [reply src/functions-reference/complex-valued_basic_functions.Rmd:65]: There is a space in every other function definition, e.g.

https://github.com/stan-dev/docs/blob/c531196bca5e02fffec822b3e7128b454cef7519/src/functions-reference/integer-valued_basic_functions.Rmd#L86


Without the spaces here, things like `operator+` were showing up twice in the index. I think the latex index functionality is pretty specific when it comes to matches

File: src/functions-reference/unbounded_continuous_distributions.Rmd
Comment (WardBrian) [reply src/functions-reference/unbounded_continuous_distributions.Rmd:818]: This also only shows up in the diff due to whitespace removal. I agree with your suggestions, but they seem best in a separate PR

**Comment by bob-carpenter** [comment]
I'm afraid merging the complex return types caused a conflict.  It should be good to go after that.

**Pull request merged**
